src/main/java/com/salesforce/dataloader/action/visitor/BulkV2Connection.java

```diff
@@ -0,0 +1,364 @@
+        }
+        connection.setRequestProperty(AUTH_HEADER, this.authHeaderValue);
+        // connection.setRequestProperty(SESSION_ID, getConfig().getSessionId());
+
+        boolean success = true;
+        InputStream in;
+        try {
+            in = connection.getInputStream();
+        } catch (IOException e) {
+            success = false;
+            in = connection.getErrorStream();
+        }
+
+        String encoding = connection.getHeaderField("Content-Encoding");
+        if ("gzip".equals(encoding)) {
+            in = new GZIPInputStream(in);
+        }
+
+        if (getConfig().isTraceMessage() || getConfig().hasMessageHandlers()) {
+            byte[] bytes = FileUtil.toBytes(in);
+            in = new ByteArrayInputStream(bytes);
+
+            if (getConfig().hasMessageHandlers()) {
+                Iterator<MessageHandler> it = getConfig().getMessagerHandlers();
+                while (it.hasNext()) {
+                    MessageHandler handler = it.next();
+                    if (handler instanceof MessageHandlerWithHeaders) {
+                        ((MessageHandlerWithHeaders)handler).handleRequest(url, new byte[0], null);
+                        ((MessageHandlerWithHeaders)handler).handleResponse(url, bytes, connection.getHeaderFields());
+                    } else {
+                        handler.handleRequest(url, new byte[0]);
+                        handler.handleResponse(url, bytes);
+                    }
+                }
+            }
+
+            if (getConfig().isTraceMessage()) {
+                getConfig().getTraceStream().println(url.toExternalForm());
+
+                Map<String, List<String>> headers = connection.getHeaderFields();
+                for (Map.Entry<String, List<String>>entry : headers.entrySet()) {
+                    StringBuffer sb = new StringBuffer();
+                    List<String> values = entry.getValue();
+
+                    if (values != null) {
+                        for (String v : values) {
+                            sb.append(v);
+                        }
+                    }
+
+                    getConfig().getTraceStream().println(entry.getKey() + ": " + sb.toString());
+                }
+
+                getConfig().teeInputStream(bytes);
+            }
+        }
+
+        if (!success) {
+            ContentType type = null;
+            String contentTypeHeader = connection.getContentType();
+            if (contentTypeHeader != null) {
+                if (contentTypeHeader.contains(XML_CONTENT_TYPE)) {
+                    type = ContentType.XML;
+                } else if (contentTypeHeader.contains(JSON_CONTENT_TYPE)) {
+                    type = ContentType.JSON;
+                }
+            }
+            parseAndThrowException(in, type);
+        }
+
+        this.queryLocator = connection.getHeaderField("Sforce-Locator");
+        this.queryResultNumberOfRecords = connection.getHeaderField("Sforce-NumberOfRecords");
```

Comment:
This method seems fairly generic and usable for different GET requests. But these last two lines are specific to GET requests for query results. Can these lines be moved somewhere else?

Reply:
Agreed. I have split the method and its usage as follows:

    
```
    private InputStream doGetQueryResultStream(URL resultsURL) throws IOException, AsyncApiException {
        HttpURLConnection httpConnection = openHttpConnection(resultsURL);
        InputStream is = doHttpGet(httpConnection, resultsURL);
        this.queryLocator = httpConnection.getHeaderField("Sforce-Locator");
        this.numberOfRecordsInQueryResult = Integer.valueOf(httpConnection.getHeaderField("Sforce-NumberOfRecords"));
        return is;
    }
    
    private InputStream doGetJobStatusStream(URL jobStatusURL) throws IOException, AsyncApiException {
        HttpURLConnection httpConnection = openHttpConnection(jobStatusURL);
        return doHttpGet(httpConnection, jobStatusURL);
    }
    
    private HttpURLConnection openHttpConnection(URL url) throws IOException {
        HttpURLConnection connection = getConfig().createConnection(url, null);
        SSLContext sslContext = getConfig().getSslContext();
        if (sslContext != null && connection instanceof HttpsURLConnection) {
            ((HttpsURLConnection)connection).setSSLSocketFactory(sslContext.getSocketFactory());
        }
        connection.setRequestProperty(AUTH_HEADER, this.authHeaderValue);
        return connection;
    }
    
    private InputStream doHttpGet(HttpURLConnection connection, URL url) throws IOException, AsyncApiException {
        boolean success = true;
        InputStream in;
        try {
            ...
            ...
```

Reply:
Yes, I have refactored doHttpGet() to be able to get the handle to HttpURLConnection object required to get the header fields and moved the 2 lines to the calling method.